test_runner/regress/test_gc_aggressive.py

```diff
@@ -175,3 +176,122 @@ def get_num_remote_ops(file_kind: str, op_kind: str) -> int:
+
+    neon_env_builder.enable_remote_storage(
+        remote_storage_kind=RemoteStorageKind.LOCAL_FS,
+        test_name="test_gc_of_remote_layers",
+    )
+
+    env = neon_env_builder.init_start()
+
+    tenant_id, timeline_id = env.neon_cli.create_tenant(
+        conf={
+            "pitr_interval": "0s",
+            "gc_period": "0s",
+            "compaction_period": "0s",
+            "compaction_threshold": "1",  # compact every time
+            "image_creation_threshold": "3",  # aggressive image creation => more layers GC'd
+            "checkpoint_timeout": "24h",  # something we won't reach
+            "checkpoint_distance": f"{1024**4}",  # something we won't reach, we checkpoint manually
+            "compaction_target_size": f"{1* 1024**2}",  # small size to maximize chance that a layer only contains obsolete keys
+        },
+    )
+    env.initial_tenant = tenant_id  # update_and_gc relies on this
+    pg = env.postgres.create_start("main")
+
+    ps_http = env.pageserver.http_client()
+
+    with pg.cursor() as cur:
+        cur.execute("CREATE TABLE tbl(id numeric, data text)")
+        cur.execute("CREATE INDEX ON tbl (id)")
+        ps_http.timeline_checkpoint(tenant_id, timeline_id)
+
+        for _ in range(0, 2):
+            layers_before = list(env.timeline_dir(tenant_id, timeline_id).glob("*-*_*"))
+
+            cur.execute("DELETE FROM tbl")
+            cur.execute("TRUNCATE TABLE tbl")  # create tombstones in the layer
+            # fill in new data, hopefully overwriting those tombstones
+            cur.execute(
+                """INSERT INTO tbl SELECT
+                    g,
+                    g || 'long string to consume some space'
+                 from generate_series(1,100) g
+                """
+            )
+            cur.execute("VACUUM tbl")
+
+            wait_for_sk_commit_lsn_to_arrive_at_pageserver_last_record_lsn(
+                tenant_id,
+                timeline_id,
+                env.safekeepers,
+                env.pageserver,
+            )
+
+            # Checkpoint and compact
+            ps_http.timeline_checkpoint(tenant_id, timeline_id)
+
+            layers_after = list(env.timeline_dir(tenant_id, timeline_id).glob("*-*_*"))
+            # ensure this loop helps grow the layer count
+            log.info("before: %s, after: %s", len(layers_before), len(layers_after))
+            assert len(layers_after) > len(layers_before)
+
+    log.info("ensure the loop above produced image and delta layers")
+    pre_evict_info = ps_http.layer_map_info(tenant_id, timeline_id)
+    log.info("layer map dump: %s", pre_evict_info)
+    log.info("by kind: %s", pre_evict_info.kind_count())
+    resident_layers = list(env.timeline_dir(tenant_id, timeline_id).glob("*-*_*"))
+    log.info("resident layers count before eviction: %s", len(resident_layers))
+
+    log.info("evict all layers")
+    ps_http.evict_all_layers(tenant_id, timeline_id)
```

Comment:
Should this eviction be moved down, after line 282?

Reply:
I think the problem was that I copy-pasted the 'resident physical size must be zero' checks to the 'remote physical size' checks.
Pushed commit [fixup: copy-pasta-error in regression test](https://github.com/neondatabase/neon/pull/3729/commits/8a12388b669e721758ded2fbc41d3a4b4c41ad06) to fix this.